worlds/osrs/__init__.py

```diff
@@ -0,0 +1,640 @@
+            task = general_tasks.pop()
+            self.add_location(task)
+            locations_added += 1
+
+        general_weight = self.options.general_task_weight if len(general_tasks) > 0 else 0
+        combat_tasks = [task for task in self.locations_by_category["combat"]
+                        if task.skills[0].level <= int(self.options.max_combat_level)]
+        # if not self.options.progressive_tasks:
+        rnd.shuffle(combat_tasks)
+        if not generation_is_fake:
+            combat_tasks = combat_tasks[0:self.options.max_combat_tasks]
+        combat_weight = self.options.combat_task_weight if len(combat_tasks) > 0 else 0
+
+        tasks_per_task_type: typing.Dict[str, typing.List[LocationRow]] = {}
+        weights_per_task_type: typing.Dict[str, int] = {}
+
+        task_types = ["prayer", "magic", "runecraft", "mining",
+                      "smithing", "fishing", "cooking", "firemaking", "woodcutting"]
+        for task_type in task_types:
+            max_level_for_task_type = getattr(self.options, f"max_{task_type}_level")
+            max_amount_for_task_type = getattr(self.options, f"max_{task_type}_tasks")
+            tasks_for_this_type = [task for task in self.locations_by_category[task_type]
+                                   if task.skills[0].level <= max_level_for_task_type]
+            if not self.options.progressive_tasks:
+                rnd.shuffle(tasks_for_this_type)
+            else:
+                tasks_for_this_type.reverse()
+
+            tasks_for_this_type = tasks_for_this_type[:max_amount_for_task_type]
+
+            tasks_per_task_type[task_type] = tasks_for_this_type
+            weights_per_task_type[task_type] = getattr(self.options,
+                                                       f"{task_type}_task_weight") if tasks_for_this_type else 0
+
+        # Build a list of collections and weights in a matching order for rnd.choices later
+        all_tasks = []
+        all_weights = []
+        for task_type in task_types:
+            all_tasks.append(tasks_per_task_type[task_type])
+            all_weights.append(weights_per_task_type[task_type])
+
+        while locations_added < locations_required or (generation_is_fake and len(all_tasks) > 0):
+            if all_tasks:
+                chosen_task = rnd.choices(all_tasks, all_weights)[0]
+                if chosen_task:
+                    task = chosen_task.pop()
+                    self.add_location(task)
+                    locations_added += 1
+
+                # This isn't an else because chosen_task can become empty in the process of resolving the above block
+                # We still want to clear this list out while we're doing that
+                if not chosen_task:
+                    index = all_tasks.index(chosen_task)
+                    del all_tasks[index]
+                    del all_weights[index]
+
+            else:
+                assert len(general_tasks) > 0, \
```

Comment:
This should probably be an exception and not an assertion, as assertions don't happen in frozen / the releases of AP